File: typescript/patterns/retry-logic.ts
/**
 * Pattern: Retry Logic with Exponential Backoff
 *
 * Demonstrates robust error handling and retry patterns for AGIRAILS requests.
 *
 * Production services need to handle:
 * - Network failures
 * - Provider timeouts
 * - Temporary unavailability
 * - Rate limiting
 *
 * This example shows:
 * - Exponential backoff with jitter
 * - Retry classification (retryable vs non-retryable errors)
 * - Circuit breaker pattern
 * - Comprehensive error handling
 *
 * Run: npx ts-node patterns/retry-logic.ts
 */

import { provide, request } from '@agirails/sdk';
import { waitForProvider } from '../src/utils/helpers';

// =========================================================
// CONFIGURATION
// =========================================================

interface RetryConfig {
  maxRetries: number;
  baseDelayMs: number;
  maxDelayMs: number;
  jitterFactor: number; // 0-1, adds randomness to prevent thundering herd
}

const DEFAULT_RETRY_CONFIG: RetryConfig = {
  maxRetries: 3,
  baseDelayMs: 1000,
  maxDelayMs: 30000,
  jitterFactor: 0.3,
};

// =========================================================
// ERROR CLASSIFICATION
// =========================================================

/**
 * Determines if an error is retryable
 */
function isRetryableError(error: any): boolean {
  const errorMessage = String(error.message || error);

  // Network errors - always retry
  if (
    errorMessage.includes('ECONNREFUSED') ||
    errorMessage.includes('ETIMEDOUT') ||
    errorMessage.includes('ENOTFOUND') ||
    errorMessage.includes('network')
  ) {
    return true;
  }

  // Timeout errors - retry
  if (errorMessage.includes('timeout') || errorMessage.includes('timed out')) {
    return true;
  }

  // Rate limiting - retry with backoff
  if (
    errorMessage.includes('rate limit') ||
    errorMessage.includes('429') ||
    errorMessage.includes('too many requests')
  ) {
    return true;
  }

  // Provider temporarily unavailable - retry
  if (
    errorMessage.includes('service unavailable') ||
    errorMessage.includes('503') ||
    errorMessage.includes('temporarily')
  ) {
    return true;
  }

  // Non-retryable errors
  if (
    errorMessage.includes('invalid input') ||
    errorMessage.includes('unauthorized') ||
    errorMessage.includes('insufficient funds') ||
    errorMessage.includes('budget exceeded')
  ) {
    return false;
  }

  // Default: don't retry unknown errors
  return false;
}

// =========================================================
// RETRY UTILITIES
// =========================================================

/**
 * Calculate delay with exponential backoff and jitter
 */
function calculateDelay(attempt: number, config: RetryConfig): number {
  // Exponential backoff: 2^attempt * baseDelay
  const exponentialDelay = Math.pow(2, attempt) * config.baseDelayMs;

  // Cap at max delay
  const cappedDelay = Math.min(exponentialDelay, config.maxDelayMs);

  // Add jitter to prevent thundering herd
  const jitter = cappedDelay * config.jitterFactor * Math.random();

  return cappedDelay + jitter;
}

/**
 * Sleep for specified milliseconds
 */
function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// =========================================================
// CORE RETRY WRAPPER
// =========================================================

interface RetryResult<T> {
  success: boolean;
  result?: T;
  error?: Error;
  attempts: number;
  totalTimeMs: number;
}

/**
 * Execute a function with retry logic
 */
async function withRetry<T>(
  fn: () => Promise<T>,
  config: RetryConfig = DEFAULT_RETRY_CONFIG,
  operationName: string = 'operation'
): Promise<RetryResult<T>> {
  const startTime = Date.now();
  let lastError: Error | undefined;

  for (let attempt = 0; attempt <= config.maxRetries; attempt++) {
    try {
      if (attempt > 0) {
        console.log(`  [Retry] Attempt ${attempt + 1}/${config.maxRetries + 1} for ${operationName}`);
      }

      const result = await fn();

      return {
        success: true,
        result,
        attempts: attempt + 1,
        totalTimeMs: Date.now() - startTime,
      };
    } catch (error: any) {
      lastError = error;

      console.log(`  [Error] ${operationName}: ${error.message}`);

      // Check if error is retryable
      if (!isRetryableError(error)) {
        console.log(`  [Retry] Error is not retryable, giving up`);
        break;
      }

      // Check if we have retries left
      if (attempt >= config.maxRetries) {
        console.log(`  [Retry] Max retries (${config.maxRetries}) exceeded`);
        break;
      }

      // Calculate and apply delay
      const delay = calculateDelay(attempt, config);
      console.log(`  [Retry] Waiting ${Math.round(delay)}ms before retry...`);
      await sleep(delay);
    }
  }

  return {
    success: false,
    error: lastError,
    attempts: config.maxRetries + 1,
    totalTimeMs: Date.now() - startTime,
  };
}

// =========================================================
// AGIRAILS-SPECIFIC RETRY WRAPPER
// =========================================================

interface AGIRAILSRequestOptions {
  service: string;
  input: any;
  budget: number;
  retryConfig?: RetryConfig;
}

/**
 * Make an AGIRAILS request with automatic retry
 */
async function requestWithRetry(options: AGIRAILSRequestOptions) {
  const config = options.retryConfig || DEFAULT_RETRY_CONFIG;

  console.log(`[Request] Starting ${options.service} request with retry...`);

  const result = await withRetry(
    async () => {
      const { result, transaction } = await request(options.service, {
        input: options.input,
        budget: options.budget,
      });
      return { result, transaction };
    },
    config,
    options.service
  );

  if (result.success) {
    console.log(`[Request] Success after ${result.attempts} attempt(s) (${result.totalTimeMs}ms)`);
  } else {
    console.log(`[Request] Failed after ${result.attempts} attempts (${result.totalTimeMs}ms)`);
  }

  return result;
}

// =========================================================
// CIRCUIT BREAKER PATTERN
// =========================================================

interface CircuitBreakerConfig {
  failureThreshold: number; // Failures before opening circuit
  resetTimeMs: number; // Time before trying again
}

class CircuitBreaker {
  private failures: number = 0;
  private lastFailureTime: number = 0;
  private isOpen: boolean = false;

  constructor(
    private name: string,
    private config: CircuitBreakerConfig
  ) {}

  async execute<T>(fn: () => Promise<T>): Promise<T> {
    // Check if circuit is open
    if (this.isOpen) {
      const timeSinceFailure = Date.now() - this.lastFailureTime;

      if (timeSinceFailure < this.config.resetTimeMs) {
        throw new Error(`Circuit breaker '${this.name}' is open. Try again in ${Math.round((this.config.resetTimeMs - timeSinceFailure) / 1000)}s`);
      }

      // Try to close circuit (half-open state)
      console.log(`[CircuitBreaker] ${this.name}: Attempting reset (half-open)`);
    }

    try {
      const result = await fn();

      // Success: reset circuit
      this.failures = 0;
      this.isOpen = false;
      return result;
    } catch (error) {
      this.failures++;
      this.lastFailureTime = Date.now();

      if (this.failures >= this.config.failureThreshold) {
        this.isOpen = true;
        console.log(`[CircuitBreaker] ${this.name}: Circuit OPEN after ${this.failures} failures`);
      }

      throw error;
    }
  }

  getStatus(): string {
    return this.isOpen ? 'open' : 'closed';
  }
}

// =========================================================
// DEMO PROVIDERS
// =========================================================

async function startDemoProviders() {
  // Unreliable provider (fails sometimes)
  let requestCount = 0;
  const unreliableProvider = provide('unreliable-service', async (job) => {
    requestCount++;
    console.log(`[Unreliable] Request #${requestCount}`);

    // Fail 60% of the time for first 3 requests
    if (requestCount <= 3 && Math.random() < 0.6) {
      throw new Error('service unavailable - temporary failure');
    }

    await new Promise((r) => setTimeout(r, 200));
    return { message: 'Success!', requestNumber: requestCount };
  });

  // Reliable provider
  const reliableProvider = provide('reliable-service', async (job) => {
    await new Promise((r) => setTimeout(r, 100));
    return { status: 'ok', input: job.input };
  });

  await Promise.all([waitForProvider(unreliableProvider), waitForProvider(reliableProvider)]);

  return {
    stop: async () => {
      await unreliableProvider.stop();
      await reliableProvider.stop();
    },
  };
}

// =========================================================
// DEMO
// =========================================================

async function main() {
  console.log('='.repeat(60));
  console.log('AGIRAILS Pattern: Retry Logic');
  console.log('='.repeat(60));
  console.log('');

  const providers = await startDemoProviders();

  try {
    // Demo 1: Request with retry
    console.log('--- Demo 1: Request with Retry ---\n');

    const result1 = await requestWithRetry({
      service: 'unreliable-service',
      input: { test: 'data' },
      budget: 1.0,
      retryConfig: {
        maxRetries: 5,
        baseDelayMs: 500,
        maxDelayMs: 5000,
        jitterFactor: 0.2,
      },
    });

    if (result1.success) {
      console.log('Result:', result1.result);
    }
    console.log('');

    // Demo 2: Circuit Breaker
    console.log('--- Demo 2: Circuit Breaker ---\n');

    const breaker = new CircuitBreaker('unreliable', {
      failureThreshold: 3,
      resetTimeMs: 5000,
    });

    for (let i = 0; i < 5; i++) {
      try {
        const result = await breaker.execute(async () => {
          // Simulate 70% failure rate
          if (Math.random() < 0.7) {
            throw new Error('Service error');
          }
          return { success: true, attempt: i + 1 };
        });
        console.log(`Attempt ${i + 1}: Success`, result);
      } catch (error: any) {
        console.log(`Attempt ${i + 1}: Failed - ${error.message}`);
      }

      await sleep(500);
    }
    console.log('');

    // Demo 3: Reliable service (no retries needed)
    console.log('--- Demo 3: Reliable Service ---\n');

    const result3 = await requestWithRetry({
      service: 'reliable-service',
      input: { message: 'Hello!' },
      budget: 0.5,
    });

    console.log('Result:', result3.result);
    console.log('');

    // Summary
    console.log('='.repeat(60));
    console.log('Summary: Retry Patterns');
    console.log('='.repeat(60));
    console.log('');
    console.log('1. withRetry(): Generic retry wrapper with exponential backoff');
    console.log('2. requestWithRetry(): AGIRAILS-specific retry wrapper');
    console.log('3. CircuitBreaker: Prevents cascading failures');
    console.log('4. isRetryableError(): Classify errors for retry decisions');
    console.log('');
    console.log('Best Practices:');
    console.log('  - Always use exponential backoff (not fixed delays)');
    console.log('  - Add jitter to prevent thundering herd');
    console.log('  - Classify errors: some are retryable, some are not');
    console.log('  - Use circuit breakers for external dependencies');
    console.log('  - Log retry attempts for debugging');
    console.log('');
  } finally {
    await providers.stop();
  }
}

main().catch(console.error);

File: python/patterns/retry_logic.py
#!/usr/bin/env python3
"""
Pattern: Retry Logic

Demonstrates resilient request handling with retries, backoff, and circuit breakers.

This pattern shows:
- Exponential backoff with jitter
- Error classification (retryable vs non-retryable)
- Circuit breaker pattern
- Request wrapping

Run: python patterns/retry_logic.py
"""

import asyncio
import sys
import random
import time
from pathlib import Path
from dataclasses import dataclass, field
from typing import TypeVar, Callable, Awaitable, Optional, List
from enum import Enum

sys.path.insert(0, str(Path(__file__).parent.parent))

from src.utils.helpers import clear_mock_state, log, log_section

T = TypeVar("T")


class CircuitState(Enum):
    CLOSED = "closed"  # Normal operation
    OPEN = "open"  # Failing, reject requests
    HALF_OPEN = "half_open"  # Testing if service recovered


@dataclass
class RetryConfig:
    """Configuration for retry behavior."""
    max_attempts: int = 3
    initial_delay_ms: int = 100
    max_delay_ms: int = 10000
    backoff_multiplier: float = 2.0
    jitter: bool = True


@dataclass
class CircuitBreakerConfig:
    """Configuration for circuit breaker."""
    failure_threshold: int = 5
    success_threshold: int = 2
    timeout_ms: int = 30000


class RetryableError(Exception):
    """Error that can be retried."""
    pass


class NonRetryableError(Exception):
    """Error that should not be retried."""
    pass


def is_retryable_error(error: Exception) -> bool:
    """Classify if an error is retryable."""
    # Network errors are retryable
    if isinstance(error, (TimeoutError, ConnectionError, RetryableError)):
        return True

    # Check error message for transient patterns
    message = str(error).lower()
    transient_patterns = [
        "timeout",
        "connection",
        "temporary",
        "unavailable",
        "rate limit",
        "too many requests",
        "503",
        "502",
    ]

    return any(pattern in message for pattern in transient_patterns)


class CircuitBreaker:
    """Circuit breaker to prevent cascading failures."""

    def __init__(self, config: CircuitBreakerConfig):
        self.config = config
        self.state = CircuitState.CLOSED
        self.failure_count = 0
        self.success_count = 0
        self.last_failure_time: Optional[float] = None

    def can_execute(self) -> bool:
        """Check if request can proceed."""
        if self.state == CircuitState.CLOSED:
            return True

        if self.state == CircuitState.OPEN:
            # Check if timeout has passed
            if self.last_failure_time:
                elapsed = (time.time() - self.last_failure_time) * 1000
                if elapsed >= self.config.timeout_ms:
                    self.state = CircuitState.HALF_OPEN
                    return True
            return False

        # HALF_OPEN - allow one request
        return True

    def record_success(self) -> None:
        """Record a successful request."""
        if self.state == CircuitState.HALF_OPEN:
            self.success_count += 1
            if self.success_count >= self.config.success_threshold:
                self.state = CircuitState.CLOSED
                self.failure_count = 0
                self.success_count = 0

    def record_failure(self) -> None:
        """Record a failed request."""
        self.failure_count += 1
        self.last_failure_time = time.time()

        if self.state == CircuitState.HALF_OPEN:
            self.state = CircuitState.OPEN
            self.success_count = 0
        elif self.failure_count >= self.config.failure_threshold:
            self.state = CircuitState.OPEN


async def with_retry(
    fn: Callable[[], Awaitable[T]],
    config: RetryConfig = RetryConfig(),
    circuit_breaker: Optional[CircuitBreaker] = None,
) -> T:
    """Execute function with retry logic."""
    last_error: Optional[Exception] = None

    for attempt in range(1, config.max_attempts + 1):
        # Check circuit breaker
        if circuit_breaker and not circuit_breaker.can_execute():
            raise Exception(f"Circuit breaker is OPEN - requests blocked")

        try:
            result = await fn()

            # Record success
            if circuit_breaker:
                circuit_breaker.record_success()

            return result

        except Exception as e:
            last_error = e

            # Record failure
            if circuit_breaker:
                circuit_breaker.record_failure()

            # Check if retryable
            if not is_retryable_error(e):
                raise e

            # Check if we have attempts left
            if attempt >= config.max_attempts:
                break

            # Calculate delay with exponential backoff
            delay_ms = config.initial_delay_ms * (config.backoff_multiplier ** (attempt - 1))
            delay_ms = min(delay_ms, config.max_delay_ms)

            # Add jitter
            if config.jitter:
                delay_ms = delay_ms * (0.5 + random.random())

            print(f"   Attempt {attempt} failed, retrying in {delay_ms:.0f}ms...")
            await asyncio.sleep(delay_ms / 1000)

    raise last_error or Exception("All retry attempts failed")


async def main() -> None:
    log_section("AGIRAILS Pattern - Retry Logic")

    clear_mock_state()

    # =====================================================
    # Part 1: Basic Retry
    # =====================================================
    log_section("Part 1: Basic Retry")

    attempt_count = 0

    async def flaky_operation() -> str:
        """Simulates an operation that fails twice then succeeds."""
        nonlocal attempt_count
        attempt_count += 1

        if attempt_count < 3:
            raise RetryableError(f"Temporary failure (attempt {attempt_count})")

        return "Success!"

    log("🔄", "Calling flaky operation with retries...")

    try:
        result = await with_retry(flaky_operation)
        print(f"   Result: {result}")
        print(f"   Total attempts: {attempt_count}")
    except Exception as e:
        print(f"   Failed: {e}")

    # =====================================================
    # Part 2: Non-Retryable Errors
    # =====================================================
    log_section("Part 2: Non-Retryable Errors")

    async def auth_error_operation() -> str:
        """Simulates an authentication error (non-retryable)."""
        raise NonRetryableError("Invalid API key")

    log("🚫", "Calling operation that fails with auth error...")

    try:
        await with_retry(auth_error_operation)
    except NonRetryableError as e:
        print(f"   Immediately failed (no retry): {e}")
    except Exception as e:
        print(f"   Error: {e}")

    # =====================================================
    # Part 3: Backoff Strategies
    # =====================================================
    log_section("Part 3: Backoff Strategies")

    print("Exponential Backoff Formula:")
    print("  delay = min(initial_delay * (multiplier ^ attempt), max_delay)")
    print()
    print("Example with default config:")
    config = RetryConfig()
    for attempt in range(1, 6):
        delay = config.initial_delay_ms * (config.backoff_multiplier ** (attempt - 1))
        delay = min(delay, config.max_delay_ms)
        print(f"  Attempt {attempt}: {delay:.0f}ms delay")
    print()

    print("With jitter (randomization):")
    print("  delay = delay * random(0.5, 1.5)")
    print("  Prevents thundering herd when many clients retry simultaneously")

    # =====================================================
    # Part 4: Circuit Breaker
    # =====================================================
    log_section("Part 4: Circuit Breaker")

    cb_config = CircuitBreakerConfig(
        failure_threshold=3,
        success_threshold=2,
        timeout_ms=5000,
    )
    circuit = CircuitBreaker(cb_config)

    failure_count = 0

    async def failing_service() -> str:
        nonlocal failure_count
        failure_count += 1
        raise RetryableError("Service unavailable")

    log("⚡", "Simulating repeated failures to trigger circuit breaker...")

    for i in range(5):
        try:
            await with_retry(
                failing_service,
                config=RetryConfig(max_attempts=1),
                circuit_breaker=circuit,
            )
        except Exception as e:
            status = circuit.state.value.upper()
            print(f"   Request {i + 1}: {str(e)[:30]}... [Circuit: {status}]")

        if circuit.state == CircuitState.OPEN:
            print(f"   Circuit OPEN - blocking further requests")
            break

    print()
    print("Circuit breaker states:")
    print("  CLOSED: Normal operation, requests flow through")
    print("  OPEN: Too many failures, requests blocked")
    print("  HALF_OPEN: Testing recovery, limited requests allowed")

    # =====================================================
    # Part 5: Error Classification
    # =====================================================
    log_section("Part 5: Error Classification")

    test_errors = [
        TimeoutError("Connection timed out"),
        ConnectionError("Cannot connect to server"),
        ValueError("Invalid input format"),
        Exception("503 Service Unavailable"),
        Exception("Rate limit exceeded"),
        Exception("Invalid signature"),
    ]

    print("Error classification:")
    for error in test_errors:
        retryable = is_retryable_error(error)
        status = "✅ Retryable" if retryable else "❌ Not retryable"
        print(f"  {status}: {type(error).__name__}: {str(error)[:30]}")

    # =====================================================
    # Part 6: Integration Example
    # =====================================================
    log_section("Part 6: Integration Example")

    print("Usage with AGIRAILS SDK:")
    print()
    print("```python")
    print("from patterns.retry_logic import with_retry, RetryConfig, CircuitBreaker")
    print()
    print("# Configure retry behavior")
    print("config = RetryConfig(")
    print("    max_attempts=5,")
    print("    initial_delay_ms=200,")
    print("    max_delay_ms=30000,")
    print("    jitter=True,")
    print(")")
    print()
    print("# Optional: circuit breaker for service-level protection")
    print("circuit = CircuitBreaker(CircuitBreakerConfig(failure_threshold=10))")
    print()
    print("# Wrap SDK calls")
    print("result = await with_retry(")
    print("    lambda: request('translate', input={'text': 'Hello'}, budget=1.0),")
    print("    config=config,")
    print("    circuit_breaker=circuit,")
    print(")")
    print("```")

    log("🎉", "Retry logic demo complete!")


if __name__ == "__main__":
    asyncio.run(main())
